src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -31,34 +37,319 @@ class CoroutineStacksPanel(project: Project) : JBPanelWithEmptyText() {
+        }
+
+    }
+
+    private fun addCoroutineInfoToCoroutineStackWindow(
+        rootNode: TreeNode<CoroutineStacksNode>,
+        rows: MutableList<Box>
+    ) {
+        val stack = Stack<Pair<TreeNode<CoroutineStacksNode>, Int>>()
+        stack.push(rootNode to 0)
+
+        while (!stack.isEmpty()) {
+            val (node, level) = stack.pop()
+
+            if (node.value.stackTrace.isNotEmpty()) {
+                addCoroutineInfoBox(node.value, rows[level])
+            }
+
+            for (child in node.children) {
+                stack.push(child to level + 1)
+            }
+        }
     }
+
+    private fun addCoroutineInfoBox(
+        value: CoroutineStacksNode,
+        box: Box
+    ) {
+
+        val headerText = CoroutineStacksBundle.message("number.of.coroutines", value.additionalData.size)
+        val stackFrames = mutableListOf<String>()
+        stackFrames.add(headerText)
+
+        for (i in value.stackTrace) {
+            i?.let { stackFrames.add(it) }
+        }
+
+        val coroutineListView = JBList<String>(stackFrames)
+
+        if (cellRenderer != null)
+            coroutineListView.cellRenderer = cellRenderer
+
+        val scrollPane = JBScrollPane(coroutineListView)
+
+        val border = BorderFactory.createLineBorder(Color.BLACK, Constants.borderWidth)
+        scrollPane.border = border
+
+        box.add(scrollPane)
+        box.add(Box.createHorizontalStrut(Constants.boxHorizontalStruct))
+
+    }
+
+    private val cellRenderer = object : DefaultListCellRenderer() {
+        private val ITEM_BORDER = BorderFactory.createMatteBorder(0, 0, 1, 0, Color.GRAY)
+
+        override fun getListCellRendererComponent(
+            list: JList<*>,
+            value: Any,
+            index: Int,
+            isSelected: Boolean,
+            cellHasFocus: Boolean
+        ): Component {
+            val renderer = super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
+            if (index < list.model.size - 1) {
+                (renderer as? JComponent)?.border = ITEM_BORDER
+            } else {
+                (renderer as? JComponent)?.border = null
+            }
+            return renderer
+        }
+    }
+
+    private fun generateParallelStackTree(
+        tree: CoroutineStacksPanel.Tree<CoroutineStacksPanel.CoroutineStacksNode>,
+        rootValue: CoroutineStacksPanel.CoroutineStacksNode?,
+        positionOfStackFrame: Int
+    ) {
+
+        val dataToStackFrame = mutableMapOf<CoroutineInfoData, String>()
+        if (rootValue == null) {
+            return
+        }
+            for (data in rootValue.additionalData) {
+                if (data.stackTrace.size > positionOfStackFrame)
+                    dataToStackFrame[data] = data.stackTrace[data.stackTrace.size -1 - positionOfStackFrame].toString()
+            }
+
+        val groupedPositions = dataToStackFrame.entries
```

Comment:
This block of code is still very cryptic, let's extract `gropBy` and `map` calls to different variables